### PR TITLE
Add the sensor model

### PR DESCRIPTION
**Pull request by nomis52**

### CODE REVIEW

File: firmware/src/rdm_responder.c [hunk 1]
```diff
@@ -37,9 +37,61 @@ const char MANUFACTURER_LABEL[] = "Open Lighting Project";
...
 RDMResponder g_responder;
 
+/*
+ * @brief Record the sensor at the specified index.
+ */
+static inline void RecordSensor(unsigned int i) {
+  if (g_responder.def->sensors[i].recorded_value_support &
+      SENSOR_SUPPORTS_RECORDING_MASK) {
+    g_responder.sensors[i].recorded_value =
+      g_responder.sensors[i].present_value;
```
Comment (peternewman) [inline firmware/src/rdm_responder.c]: Indent four from the line above.

File: firmware/src/rdm_responder.c [hunk 2]
```diff
@@ -51,18 +103,19 @@ void RDMResponder_ResetToFactoryDefaults() {
...
-            g_responder.def->default_device_label,
-            RDMUtil_SafeStringLength(g_responder.def->default_device_label,
-                                     RDM_DEFAULT_STRING_SIZE));
-    g_responder.device_label[RDM_DEFAULT_STRING_SIZE] = 0;
+    unsigned int len = RDMUtil_StringCopy(
+        g_responder.device_label,
+        RDM_DEFAULT_STRING_SIZE,
+        g_responder.def->default_device_label,
+        RDM_DEFAULT_STRING_SIZE);
+    g_responder.device_label[len] = 0;
```
Comment (peternewman) [inline firmware/src/rdm_responder.c]: It looks like this adds the null in RDMUtil_StringCopy so you don't need to now?

File: firmware/src/rdm_responder.c [hunk 3]
```diff
@@ -451,43 +503,159 @@ int RDMResponder_SetDeviceLabel(const RDMHeader *header,
...
+        SENSOR_SUPPORTS_RECORDING_MASK) {
+      RecordSensor(sensor_index);
+      return RDMResponder_BuildSetAck(header);
+    } else {
       return RDMResponder_BuildNack(header, NR_DATA_OUT_OF_RANGE);
+    }
+  } else if (sensor_index == ALL_SENSORS) {
+    unsigned int i = 0;
+    for (; i < g_responder.def->sensor_count; i++) {
+      ResetSensor(i);
```
Comment (peternewman) [inline firmware/src/rdm_responder.c]: RecordSensor

Comment (peternewman) [inline firmware/src/rdm_responder.c]: Does this also need to check that each sensor has recorded_value_support?

File: firmware/src/sensor_model.c [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+static void SensorModel_Activate() {
+  g_responder.def = &RESPONDER_DEFINITION;
+  unsigned int i = 0;
+  for (; i < NUMBER_OF_SENSORS; i++) {
+    g_sensor_model.sensors[i].present_value = 0;
+    g_sensor_model.sensors[i].lowest_value = 0;
+    g_sensor_model.sensors[i].highest_value = 0;
+    g_sensor_model.sensors[i].recorded_value = 0;
+    if (i == 1) {
+      // Sensor 1 always nacks with a HARDWARE_FAULT
```
Comment (peternewman) [inline firmware/src/sensor_model.c]: Clarify the comment regarding 0 based numbering, perhaps give the sensor name instead.

File: firmware/src/rdm_responder.c [hunk 5]
```diff
@@ -37,9 +37,61 @@ const char MANUFACTURER_LABEL[] = "Open Lighting Project";
...
+/*
+ * @brief Reset the sensor at the specified index.
+ */
+static void ResetSensor(unsigned int i) {
+  if (g_responder.def->sensors[i].recorded_value_support &
+      SENSOR_SUPPORTS_LOWEST_HIGHEST_MASK) {
+    g_responder.sensors[i].lowest_value = g_responder.sensors[i].present_value;
+    g_responder.sensors[i].highest_value = g_responder.sensors[i].present_value;
+  } else {
+    g_responder.sensors[i].lowest_value = 0;
```
Comment (peternewman) [inline firmware/src/rdm_responder.c]: Can you make the zero a constant like SENSOR_RECORDED_RANGE_UNSUPPORTED in OLA

File: firmware/src/rdm_responder.c [hunk 6]
```diff
@@ -37,9 +37,61 @@ const char MANUFACTURER_LABEL[] = "Open Lighting Project";
...
+    g_responder.sensors[i].highest_value = 0;
+  }
+
+  if (g_responder.def->sensors[i].recorded_value_support &
+      SENSOR_SUPPORTS_RECORDING_MASK) {
+    g_responder.sensors[i].recorded_value =
+      g_responder.sensors[i].present_value;
+  } else {
+    g_responder.sensors[i].recorded_value = 0;
```
Comment (peternewman) [inline firmware/src/rdm_responder.c]: Likewise SENSOR_RECORDED_UNSUPPORTED

File: firmware/src/sensor_model.c [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+void SampleSensors() {
+  g_sensor_model.sensor_sample_time = CoarseTimer_GetTime();
+
+  unsigned int i = 0;
+  for (; i < NUMBER_OF_SENSORS; i++) {
+    int16_t new_value = (Random_PseudoGet() % (
+        RESPONDER_DEFINITION.sensors[i].range_maximum_value -
+        RESPONDER_DEFINITION.sensors[i].range_minimum_value)) +
+        RESPONDER_DEFINITION.sensors[i].range_minimum_value;
+    g_sensor_model.sensors[i].present_value = new_value;
```
Comment (peternewman) [inline firmware/src/sensor_model.c]: Can you split this part out into a function, so we can implement real sensors by just calling the function.

**Comment by peternewman** [comment]
Mostly minor comments

File: firmware/src/rdm_responder.c (hunk 1 above)
Comment (nomis52) [inline firmware/src/rdm_responder.c]: Done

File: firmware/src/rdm_responder.c (hunk 5 above)
Comment (nomis52) [inline firmware/src/rdm_responder.c]: Done

File: firmware/src/rdm_responder.c (hunk 2 above)
Comment (nomis52) [inline firmware/src/rdm_responder.c]: Correct, done

File: firmware/src/rdm_responder.c (hunk 3 above)
Comment (nomis52) [inline firmware/src/rdm_responder.c]: Fixed, the checking is done in the Record sensor function

File: firmware/src/sensor_model.c (hunk 4 above)
Comment (nomis52) [inline firmware/src/sensor_model.c]: Done

File: firmware/src/sensor_model.c (hunk 7 above)
Comment (nomis52) [inline firmware/src/sensor_model.c]: Done, with tests

File: firmware/src/rdm.h [hunk 8]
```diff
@@ -122,6 +122,26 @@ enum {
...
+
+/**
+ * @brief The all-sensors value.
+ */
+static const uint8_t ALL_SENSORS = 0xff;
+
+/**
+ * @brief Indicates that recorded / highest / lowest values aren't supported.
+ */
+static const uint16_t SENSOR_VALUE_UNSUPPORTED = 0;
```
Comment (peternewman) [inline firmware/src/rdm.h]: Can we have this as two constants, like we do in OLA; I guess they're defined separately in some sense in the docs?

Comment (nomis52) [inline firmware/src/rdm.h]: It's the same value, no point in wasting memory.

Comment (peternewman) [inline firmware/src/rdm.h]: Okay. I assumed the compiler would just swap it out so there wouldn't be a memory impact.

File: firmware/src/rdm_util.c [hunk 9]
```diff
@@ -78,3 +90,15 @@ unsigned int RDMUtil_SafeStringLength(const char *str, unsigned int max_size) {
   return size;
 }
 
+void RDMUtil_UpdateSensor(SensorData *sensor, uint8_t recorded_value_support,
+                          int16_t new_value) {
+  sensor->present_value = new_value;
+  if (recorded_value_support & SENSOR_SUPPORTS_LOWEST_HIGHEST_MASK) {
+    if (new_value < sensor->lowest_value) {
```
Comment (peternewman) [inline firmware/src/rdm_util.c]: Random thought, is this more efficient than calling min and always updating the value, I guess the answer is yes, but I'm not sure.

Comment (nomis52) [inline firmware/src/rdm_util.c]: I suspect it is. You save an assignment for the negative branch.

**Pull request merged**